### ISSUE TITLE
Rate Limit Middleware

### ISSUE THREAD

**Issue by TPME** [middyjs/middy#894]
**Is your feature request related to a problem? Please describe.**
Rate limiting blocks users, bots, or applications that are over-using or abusing a web property. Rate limiting can stop certain kinds of bot attacks. This is especially useful for applications that are serverless, as a malicious actor could abuse the scalability feature of serverless applications and generate huge costs for the operator.

**Describe the solution you'd like**
A new middleware could be added that implements rate limiting based on custom rules by the user (e.g., key, limit, action). The storage could also be variable, there are multiple ideas... having it connect to a Redis instance or doing everything in-memory, see e.g. https://www.npmjs.com/package/lambda-rate-limiter

**Describe alternatives you've considered**
Basically https://www.npmjs.com/package/lambda-rate-limiter looks promising, but as I'm in love with middy, I'd prefer it to be available as a middleware. This way, more people can benefit from something that should be part of every modern API.

**Comment by willfarrell** [comment]
Thanks for your feedback. As much as I'd love the idea of having middy be a one-stop shop, I don't think a rate limiter is a good fit to be included in core

1. Rate limiting is better handled by a WAF, it's faster and at the edge.
2. Rate limiting using in-memory defeats the purpose as it's not shared across all lambdas running in parallel to handle the load.
3. If we include redis, there will be requests to include other storage providers. This create a maintainability burden. This is why we deprecated `db-manager`.

If you'd like to create your own middleware and share it with the community, I'd be happy to have it included in our list of community middlewares.
